tests/postgres_tests/test_search.py

```diff
@@ -453,3 +459,111 @@ def test_str(self):
+                'dialogue',
+                SearchQuery('cadeaux', config=F('dialogue_config')),
+                config=F('dialogue_config'),
+            ),
+        )
+        self.assertEqual(str(searched.query).count('"dialogue_config"::regconfig'), 2)
+        self.assertIn('<b>', searched.first().headline)
+
+    def test_headline_separator_options(self):
+        searched = Line.objects.filter(character=self.minstrel).annotate(
+            headline=SearchHeadline(
+                'dialogue',
+                SearchQuery('brave sir robin'),
+                options={'StartSel': '<strong class="clazz">', 'StopSel': '</strong>'},
+            ),
+        )
+        self.assertIn('<strong class="clazz">', searched.first().headline)
+        self.assertIn('</strong>', searched.first().headline)
+
+    def test_headline_explicit_config_and_options(self):
+        searched = Line.objects.filter(character=self.minstrel).annotate(
+            headline=SearchHeadline(
+                'dialogue',
+                SearchQuery('brave sir robin', config='english'),
+                options={'HighlightAll': True},
+            ),
+        )
+        self.assertIn('HighlightAll=true', str(searched.query))
+        self.assertIn('<b>', searched.first().headline)
+
+    def test_headline_config_short_word(self):
+        searched = Line.objects.filter(character=self.minstrel).annotate(
+            headline=SearchHeadline(
+                'dialogue',
+                SearchQuery('brave sir robin', config='english'),
+                options={'ShortWord': 6},
+            ),
+        )
+        self.assertIn('ShortWord=6', str(searched.query))
+        self.assertIn('<b>', searched.first().headline)
+
+    def test_headline_config_fragments(self):
+        searched = Line.objects.filter(character=self.minstrel).annotate(
+            headline=SearchHeadline(
+                'dialogue',
+                SearchQuery('brave sir robin', config='english'),
+                options={
+                    'FragmentDelimiter': "'''",
+                    'MaxFragments': 4,
+                    'MaxWords': 3,
+                    'MinWords': 1
+                },
+            ),
+        )
+        self.assertIn('FragmentDelimiter', str(searched.query))
+        self.assertIn('MaxFragments=4', str(searched.query))
+        self.assertIn('MaxWords=3', str(searched.query))
+        self.assertIn('MinWords=1', str(searched.query))
+        self.assertIn("</b>'''<b>", searched.first().headline)
+
+    def test_headline_invalid_option(self):
+        msg = "Unknown options key(s) 'InvalidOption, AnotherInvalidOption'."
+        with self.assertRaisesMessage(ValueError, msg):
+            SearchHeadline(
+                'dialogue',
+                SearchQuery('brave sir robin'),
+                options={
+                    'InvalidOption': 'foo',
+                    'AnotherInvalidOption': 'bar',
+                    'MaxFragments': 4
```

Comment:
This test is really critical because without the whitelisting injection could take via malformed keys.

Reply:
Yes it's a bit scary. Let me know if you got any ideas to make it more solid. Maybe adding a comment to not have it accidentally removed could help?